src/simulators/density_matrix/densitymatrix_state_chunk.hpp

```diff
@@ -437,32 +420,95 @@ auto State<densmat_t>::move_to_matrix()
 
-#pragma omp parallel for if(BaseState::chunk_omp_parallel_) private(iChunk)
-    for(iChunk=1;iChunk<BaseState::num_local_chunks_;iChunk++){
-      auto tmp = BaseState::qregs_[iChunk].vector();
-      uint_t j,offset = iChunk << BaseState::chunk_bits_;
-      for(j=0;j<tmp.size();j++){
-        state[offset + j] = tmp[j];
+    auto recv = BaseState::qregs_[0].copy_to_matrix();
+
+#ifdef AER_MPI
+    //gather states from other processes
+    for(iChunk=BaseState::num_local_chunks_;iChunk<BaseState::num_global_chunks_;iChunk++){
+      BaseState::recv_data(recv.data(),size,0,iChunk);
+
+      int_t i;
+      uint_t irow_chunk = ((iChunk) >> ((BaseState::num_qubits_ - BaseState::chunk_bits_))) << (BaseState::chunk_bits_);
+      uint_t icol_chunk = ((iChunk) & ((1ull << ((BaseState::num_qubits_ - BaseState::chunk_bits_)))-1)) << (BaseState::chunk_bits_);
+#pragma omp parallel for if(num_threads > 1) num_threads(num_threads)
+      for(i=0;i<size;i++){
+        uint_t irow = i >> (BaseState::chunk_bits_);
+        uint_t icol = i & mask;
+        matrix(icol_chunk+icol,irow_chunk+irow) = recv(icol,irow);
       }
     }
+#endif
 
+    for(iChunk=0;iChunk<BaseState::num_local_chunks_;iChunk++){
+      int_t i;
+      uint_t irow_chunk = ((iChunk + BaseState::global_chunk_index_) >> ((BaseState::num_qubits_ - BaseState::chunk_bits_))) << (BaseState::chunk_bits_);
+      uint_t icol_chunk = ((iChunk + BaseState::global_chunk_index_) & ((1ull << ((BaseState::num_qubits_ - BaseState::chunk_bits_)))-1)) << (BaseState::chunk_bits_);
+      if(copy){
+        auto tmp = BaseState::qregs_[iChunk].copy_to_matrix();
+#pragma omp parallel for if(num_threads > 1) num_threads(num_threads)
+        for(i=0;i<size;i++){
+          uint_t irow = i >> (BaseState::chunk_bits_);
+          uint_t icol = i & mask;
+          matrix(icol_chunk+icol,irow_chunk+irow) = tmp(icol,irow);
+        }
+      }
+      else{
+        auto tmp = BaseState::qregs_[iChunk].move_to_matrix();
+#pragma omp parallel for if(num_threads > 1) num_threads(num_threads)
+        for(i=0;i<size;i++){
+          uint_t irow = i >> (BaseState::chunk_bits_);
+          uint_t icol = i & mask;
+          matrix(icol_chunk+icol,irow_chunk+irow) = tmp(icol,irow);
+        }
+      }
+    }
+  }
+  else{
 #ifdef AER_MPI
-    BaseState::gather_state(state);
+    //send matrices to process 0
+    for(iChunk=0;iChunk<BaseState::num_global_chunks_;iChunk++){
+      uint_t iProc = BaseState::get_process_by_chunk(iChunk);
+      if(iProc == BaseState::distributed_rank_){
+        if(copy){
+          auto tmp = BaseState::qregs_[iChunk-BaseState::global_chunk_index_].copy_to_matrix();
+          BaseState::send_data(tmp.data(),size,iChunk,0);
+        }
+        else{
+          auto tmp = BaseState::qregs_[iChunk-BaseState::global_chunk_index_].move_to_matrix();
+          BaseState::send_data(tmp.data(),size,iChunk,0);
```

Comment:
Same doubt here. Do we really need to copy/move to `tmp`? Cannot we just pass internal `BaseState::qregs_[iChunk-BaseState::global_chunk_index_]` data to `send_data` function?

Reply:
For GPU, we can send data directly from GPU to GPU on the other process if GPUDirectRDMA is supported on the system, but we can not send data directly from GPU to CPU memory on the other process. Here matrix elements are gathered on CPU so we have to copy matrix to CPU before send matrix via MPI.